libbpf-tools/oomkill.c

```diff
@@ -0,0 +1,103 @@
+// SPDX-License-Identifier: (LGPL-2.1 OR BSD-2-Clause)
+// Copyright (c) 2022 Jingxiang Zeng
+//
+// Based on oomkill(8) from BCC by Brendan Gregg.
+// 13-Jan-2022   Jingxiang Zeng   Created this.
+#include <argp.h>
+#include <errno.h>
+#include <signal.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+#include <time.h>
+#include <unistd.h>
+
+#include <bpf/bpf.h>
+#include <bpf/libbpf.h>
+#include "oomkill.skel.h"
+#include "oomkill.h"
+#include "trace_helpers.h"
+
+const char *argp_program_version = "oomkill 0.1";
+const char *argp_program_bug_address = "linuszeng@tencent.com";
+
+void handle_event(void *ctx, int cpu, void *data, __u32 data_sz) {
+	FILE *f;
+	char buf[256];
+	int n;
+	struct tm *tm;
+	char ts[32];
+	time_t t;
+	struct data_t *e = data;
+
+	f = fopen("/proc/loadavg", "r");
+	if (f) {
+		memset(buf, 0 , sizeof(buf));
+		n = fread(buf, 1, sizeof(buf), f);
+		if (n) {
+			time(&t);
+			tm = localtime(&t);
+			strftime(ts, sizeof(ts), "%H:%M:%S", tm);
+			printf("%s Triggered by PID %d (\"%s\"), OOM kill of PID %d (\"%s\"), %lld pages, loadavg: %s\n",
+				ts, e->fpid, e->fcomm, e->tpid, e->tcomm, e->pages, buf);
+		}
+		fclose(f);
+	}
+}
+
+void handle_lost_events(void *ctx, int cpu, __u64 lost_cnt)
+{
+	printf("Lost %llu events on CPU #%d!\n", lost_cnt, cpu);
+}
+
+int main(int argc, char **argv) {
+	struct perf_buffer_opts pb_opts;
+	struct perf_buffer *pb = NULL;
+	struct oomkill_bpf *obj;
+	int err;
+
+	err = bump_memlock_rlimit();
+	if (err) {
+		fprintf(stderr, "failed to increase rlimit: %d\n", err);
+		return 1;
+	}
+
+	obj = oomkill_bpf__open();
+	if (!obj) {
+		fprintf(stderr, "failed to open BPF object\n");
+		return 1;
+	}
+
+	err = oomkill_bpf__load(obj);
+	if (err) {
+		fprintf(stderr, "failed to load BPF object: %d\n", err);
+		goto cleanup;
+	}
+
+	err = oomkill_bpf__attach(obj);
+	if (err) {
+		fprintf(stderr, "failed to attach BPF programs\n");
+		goto cleanup;
+	}
+
+	printf("Tracing OOM kills... Ctrl-C to stop.\n");
+	pb_opts.sample_cb = handle_event;
+	pb_opts.lost_cb = handle_lost_events;
+	pb = perf_buffer__new(bpf_map__fd(obj->maps.events), 64, &pb_opts);
```

Comment:
Use the new opts-based `perf_buffer__new` API instead.

Reply:
done